httpx/_transports/asgi.py

```diff
@@ -120,38 +192,76 @@ async def receive() -> dict:
             return {"type": "http.request", "body": body, "more_body": True}
 
         async def send(message: dict) -> None:
-            nonlocal status_code, response_headers, body_parts
-            nonlocal response_started, response_complete
+            nonlocal status_code, response_headers
 
             if message["type"] == "http.response.start":
-                assert not response_started
+                assert not response_started.is_set()
 
                 status_code = message["status"]
                 response_headers = message.get("headers", [])
-                response_started = True
+                response_started.set()
 
             elif message["type"] == "http.response.body":
                 assert not response_complete.is_set()
                 body = message.get("body", b"")
                 more_body = message.get("more_body", False)
 
                 if body and method != b"HEAD":
-                    body_parts.append(body)
+                    await produce_response_body_chunk(body)
 
                 if not more_body:
                     response_complete.set()
 
-        try:
-            await self.app(scope, receive, send)
-        except Exception:
-            if self.raise_app_exceptions or not response_complete:
-                raise
+        async def run_app() -> None:
+            nonlocal app_exception
+            try:
+                await self.app(scope, receive, send)
+            except Exception as exc:
+                app_exception = exc
+                app_crashed.set()
+
+        aclose_app = await create_background_task(run_app)
+
+        await run_until_first_complete(app_crashed.wait, response_started.wait)
```

Comment:
Hi @florimondmanca, I was looking at your PR to learn the ASGITransport mindset. I wondered how to pass `timeout` logic into  `run_until_first_complete`. I noticed that we have no option for `timeout` for this. Am I missing something?

Reply:
You're right, currently there's no `timeout` logic in the ASGI transport, but I'm considering this as out of scope for this PR - i.e. we could add it separately.